### PR TITLE
tf -> tf2

### PR DESCRIPTION
**Pull request by gaya-**
This is a part of a series of pull requests concerning the switch from TF to TF2.
All the PRs are:

roslisp_common: https://github.com/ros/roslisp_common/pull/23
cram_highlevel: https://github.com/cram-code/cram_highlevel/pull/46
cram_physics: https://github.com/cram-code/cram_physics/pull/23
cram_pr2: https://github.com/cram-code/cram_pr2/pull/10
cram_bridge: https://github.com/cram-code/cram_bridge/pull/21

There is still a problem with TF for projection as the joint state publisher from moveit interferes with the transforms published from projection. This should not affect the execution on the real robot.

One important change is to use ```transform-pose``` instead of ```ensure-pose-...``` as the latter would block infinitely if a transform would not be provided to TF. The function ```transfrom-pose``` has an argument called ```timeout``` which specifies in seconds how long to wait for the transform before giving up. The default value is specified in ```cram-roslisp-common:*tf-default-timeout*```. So, before starting your scenarios set it up for a reasonable value, e.g. 10 seconds or more when working on a real robot.

### CODE REVIEW

**Comment by fairlight1337** [comment]
I won't go for getting rid of `ensure-pose-stamped-transformed`. On real robots, it is a vital functionality to wait for transforms which aren't available right now (as real systems tend to take time out of whatever reasons).

We can add a timeout parameter there, but I'm definitely in for waiting indefinitely. When `indefinitely` happens, something else is wrong anyway. What would your reaction be if the transform couldn't be found? There are only two legit alternatives then:

  * Either, retrying to transform (resulting in indefinite time when no transform gets available), or
  * Stopping the plan and checking what went wrong (which will be independent of the plan's actual purpose, and which you'll need to do too when you're waiting for a transformation for too long).

So, why not wait for forever? Do you have a use-case where this default behavior actually gets problematic? @airballking, do you?

If you insist on not waiting for forever indefinitely, I strongly suggest adding a flag that signals *wait forever*, as this is actually something someone uses.

**Comment by gaya-** [comment]
> On real robots, it is a vital functionality to wait for transforms which aren't available right now (as real systems tend to take time out of whatever reasons).

Yes. And it's not even only the real robots, in the simulation there can also be a little delay before the necessary transform becomes available as the transform publishers are publishing with a discrete frequency. So I totally agree with you on this. That is why ```transform-pose``` can wait for transforms.

> I won't go for getting rid of ```ensure-pose-stamped-transformed```

Just to make sure that we're on the same page here, the only difference between ```ensure-pose...``` and ```transform-pose``` is that ```ensure-pose...``` does the waiting on ```roslisp``` side and ```trasform-pose``` utilizes the same waiting mechanism that is already implemented in the TF2 server. And, of course, that ```ensure-pose-...``` can get into an infinite loop and ```transform-pose``` cannot.

> When indefinitely happens, something else is wrong anyway.
> So, why not wait for forever?

As we all know, robotic systems are very prone to failures and 90% of robotics code is failure handling. So, having something else going wrong is not rare. If something is wrong the usual convention in computer science is to issue an informative error. Having your system suddenly block forever is not very informative: you don't know what exactly is blocking your system. Maybe your robot is just thinking very hard, maybe it is in an infinite Prolog loop, maybe one part of the system just died and it's waiting for it, or maybe it is waiting for a transform that will never be published. In that sense, trying to stop these kinds of infinite loops and issuing an informative error that can also be handled by the program itself, is a good practice, in my humble opinion. That is why I would prefer to have a ```timeout``` for any function with an endless loop that is expected to return.

> If you insist on not waiting for forever indefinitely, I strongly suggest adding a flag that signals wait forever, as this is actually something someone uses.

In our case, I would argue that "forever" is still a measurable quantity. You probably would not want your robot to wait for 3 days for a transform. In that sense, I guess there is a time limit above which you would say that if the transform is still not there then it is not a normal behavior but an error happened in the system. The limit on how big ```timeout``` can be is probably the biggest integer in seconds (I'll have to test that), which is about 40 years for 32 bit integers I'd guess.

Setting a flag to T or setting ```cram-roslisp-common:*tf-default-timeout*``` to ```1d10``` is not such a big difference, in my opinion. If you can't be bothered writing ```1d10``` we could have a helper function  ```set-timeout-to-infinity``` that would do that for you, what do you think? However, I would still advice you to think about a reasonable value for the timeout, say 2 mins, such that if something is actually wrong with TF that you would get a TF error and be sure that it was not any other part of the system that caused the loop. This would be especially helpful for, e.g., running pick and place for the whole night on the robot. If a TF error is issued the program could automatically restart the TF publisher to recover.

**Comment by fairlight1337** [comment]
So where do you actually use such a timeout? Or is this a purely theoretical argument?

**Comment by gaya-** [comment]
You mean the ```cram-roslisp-common:*tf-default-timeout*``` variable? Inside CRAM it is used in all the calls to ```transform-pose```, see, e.g. https://github.com/gaya-/cram_bridge/blob/master/cram_moveit/src/collision-environment.lisp#L291

I set it to 2 seconds in my "demo", you can set it to whatever you want in your packages.

**Comment by gaya-** [comment]
Per default, it is set to 0.0 as that's the default behaviour of the TF server:
https://github.com/cram-code/cram_highlevel/pull/46/files#diff-9ba5d240120f6d2c6c9acd33356be60dR35

If you want to change the default to a very big value that is probably fine with me. But then the people who are new to the system will be really confused about why the code they're executing is hanging and stuff. But the new people would need assistance anyway.

**Comment by fairlight1337** [comment]
Since we're going to break the API of TF again anyway (`tf` -> `tf2` change), I don't mind removing `ensure-pose-stamped-transformed` and changing the signature of `transform-pose` as long as it properly supports waiting for a transform indefinitely via a parameter, rather than *setting a global variable to enforce this behavior*.

That should suit your argument.

**Comment by gaya-** [comment]
Ok, that also sounds legit.

However, I would still need a parameter to specify the exact amount of time I would like to wait for a transform. Otherwise, one would have to write the same hard-coded integer in all the ```transform-pose``` calls, such as ```(transform-pose ... :timeout 2.0)```. And if at some point one decided that 2.0 is too much or too few, a huge amount of ```transform-pose``` calls would have to be refactored. That is why I need a default timeout parameter. And having a global parameter to parameterize the behaviour you would like your robot to have is not a bad programming style and we utilize it a lot in CRAM. If you don't like the idea of global variables (that is the impression I got from your last comment) then we can put it into the Prolog facts knowledge base.

To sum up, I suggest we change the implementation of ```transform-pose``` to

```lisp
(defun transform-pose (... :timeout timeout)
      (if (eq :timeout :infinity)
           wait-forever
           dont-wait-forever)))
```

Then you can write ```(transform-pose my-pose :timeout :infinity)```

That way you'll have your parameter to specify the wait forever behaviour, right?
